### PR TITLE
Core: Prevent potential NPEs when retrieving JSON fields

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/util/JsonUtil.java [hunk 1]
```diff
@@ -98,6 +98,11 @@ public static <T> T parse(String json, FromJson<T> parser) {
     }
   }
 
+  public static JsonNode get(String property, JsonNode node) {
+    Preconditions.checkArgument(node.has(property), "Cannot parse missing object '%s'", property);
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/util/JsonUtil.java:102]: Question: Why not use `hasNonNull`? If anything I'd make a method for getting null-able and non-null.

File: core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java [hunk 2]
```diff
@@ -383,7 +383,7 @@ private static MetadataUpdate readUpgradeFormatVersion(JsonNode node) {
 
   private static MetadataUpdate readAddSchema(JsonNode node) {
     Preconditions.checkArgument(node.hasNonNull(SCHEMA), "Cannot parse missing field: schema");
-    JsonNode schemaNode = node.get(SCHEMA);
+    JsonNode schemaNode = JsonUtil.get(SCHEMA, node);
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java:386]: Consider relying on the `Precondition` inside of `JsonUtil` (and have it use `hasNonNull`). That pattern seems to exist a lot still and could be pushed into `JsonUtil`.

**Review by kbendick — COMMENTED** [review]
This looks good to me, other than possibly pushing down the existing `hasNonNull` calls into `JsonUtil`. We rarely use `null` anyway in most parsers.

File: core/src/main/java/org/apache/iceberg/util/JsonUtil.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/util/JsonUtil.java:102]: that's a good point. I updated this to use `hasNonNull`

File: core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java:386]: makes total sense, updated

File: core/src/main/java/org/apache/iceberg/SchemaParser.java [hunk 3]
```diff
@@ -183,7 +183,6 @@ public static String toJson(Schema schema, boolean pretty) {
   private static Type typeFromJson(JsonNode json) {
     if (json.isTextual()) {
       return Types.fromPrimitiveString(json.asText());
-
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SchemaParser.java:186]: Nit: unnecessary whitespace change.

File: core/src/main/java/org/apache/iceberg/util/JsonUtil.java [hunk 4]
```diff
@@ -98,6 +98,12 @@ public static <T> T parse(String json, FromJson<T> parser) {
     }
   }
 
+  public static JsonNode get(String property, JsonNode node) {
+    Preconditions.checkArgument(
+        node.hasNonNull(property), "Cannot parse missing object '%s'", property);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/util/JsonUtil.java:103]: The error message should be `"Cannot parse missing field: %s"`. We generally prefer context after `:` rather than quoting values for clarity.

File: core/src/main/java/org/apache/iceberg/SchemaParser.java (hunk 3 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/SchemaParser.java:186]: yes I know but I didn't feel it's worth opening up a separate PR to remove this whitespace line

File: core/src/main/java/org/apache/iceberg/util/JsonUtil.java (hunk 4 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/util/JsonUtil.java:103]: updated

**Review by kbendick — APPROVED** [review]
Overall this looks good to me.

One thing of note is that we don't have a test suite for `JsonUtils`. Maybe this is a good time to add one for just one method or so and then we can add more tests later as time permits?

Seems like an important part of the code that deserves its own tests, though if that's out of scope for this PR, that makes sense.

File: core/src/main/java/org/apache/iceberg/SchemaParser.java (hunk 3 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/SchemaParser.java:186]: Why remove it at all? Seems like an unncessary change.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @nastra!

File: core/src/main/java/org/apache/iceberg/SchemaParser.java (hunk 3 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/SchemaParser.java:186]: I stumbled over it already a couple of times and figured I just remove this unnecessary line ;)
